# Lab 1 część druga - biblioteka pandas: przetwarzanie danych

Pandas jest biblioteką języka Python umożliwiającą przetwarzanie i wizualizację danych. Dzięki wygodnemu i przejrzystemu interfejsowi, jest jednym z najczęściej wykorzystywanych rozwiązań w analizie danych. 

## Instalacja

W celu instalacji biblioteki pandas należy użyć polecenia **pip install pandas**.

## Pierwsze kroki z biblioteką Pandas

Biblioteka pandas jest narzędziem zewnętrznym, więc w pierwszej kolejności należy ją zaimportować:

In [1]:
import pandas as pd  

### Wczytywanie danych

Dane w celu ich przeanalizowania powinny zostać wczytane z dowolnego zewnętrznego źródła. Jednym z najpopularniejszych formatów transferu ustraukturyzowanych danych są pliki csv. Można je wczytać za pomocą funkcji read_csv z pakietu pandas.

In [5]:
movie: pd.DataFrame = pd.read_csv('movie.csv')  

Dane z pliku csv zostały wczytane do obiektu klasy DataFrame, czyli ramki danych w formie tabeli. Wczytane dane można wyświetlić w łatwy sposób, przy użyciu notatnika Jupyter:

In [6]:
movie  

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
4911  Color        Scott Smith                     1.0      87.0   
4912  Color                NaN                    43.0      43.0   
4913  Color   Benjamin Roberds                    13.0      76.0   
4914  Color        Daniel Hsia                    14.0     100.0   
4915  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
4911                   637.0          NaN                     Comedy|Drama   
4912                   841.0          NaN     Crime|Drama|Mystery|Thriller   
4913                     0.0          NaN            Drama|Horror|Thriller   
4914                   946.0      10443.0             Comedy|Drama|Romance   
4915                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
4911  ...                  6.0  English   Canada             NaN          NaN   
4912  ...                359.0  English      USA           TV-14          NaN   
4913  ...                  3.0  English      USA             NaN       1400.0   
4914  ...                  9.0  English      USA           PG-13          NaN   
4915  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

Ramka danych składa się z nazwanych kolumn oraz z indeksów wierszy w zakresie od 0 do 4915.

W przypadku gdy ramka danych zawiera wiele wierszy, można wykorzystać metodę head wywołaną na obiekcie klasy DataFrame, która wyświetli tylko n pierwszych wierszy:

In [7]:
movie.head(n=3)  

color   director_name  num_critic_for_reviews  duration  \
0  Color   James Cameron                   723.0     178.0   
1  Color  Gore Verbinski                   302.0     169.0   
2  Color      Sam Mendes                   602.0     148.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  

[3 rows x 28 columns]

Analogicznie metoda tail wyświetli n ostatnich wierszy:

In [8]:
movie.tail(n=4) 

color     director_name  num_critic_for_reviews  duration  \
4912  Color               NaN                    43.0      43.0   
4913  Color  Benjamin Roberds                    13.0      76.0   
4914  Color       Daniel Hsia                    14.0     100.0   
4915  Color          Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes    gross                        genres  ...  \
4912                   841.0      NaN  Crime|Drama|Mystery|Thriller  ...   
4913                     0.0      NaN         Drama|Horror|Thriller  ...   
4914                   946.0  10443.0          Comedy|Drama|Romance  ...   
4915                    86.0  85222.0                   Documentary  ...   

     num_user_for_reviews language  country  content_rating  budget  \
4912                359.0  English      USA           TV-14     NaN   
4913                  3.0  English      USA             NaN  1400.0   
4914                  9.0  English      USA           PG-13     NaN   
4915                 84.0  English      USA              PG  1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
4912         NaN                  593.0        7.5         16.00   
4913      2013.0                    0.0        6.3           NaN   
4914      2012.0                  719.0        6.3          2.35   
4915      2004.0                   23.0        6.6          1.85   

     movie_facebook_likes  
4912                32000  
4913                   16  
4914                  660  
4915                  456  

[4 rows x 28 columns]

Plik movie.csv został wczytany za pomocą funkcji read_csv przy wskazaniu jedynie ścieżki. Funkcja read_csv udostępnia więcej opcji, które umożliwiają personalizację wczytywania danych. Przykładowo, plik csv jest klasycznym plikiem tekstowym, w którym wartości są odseparowane dowolnym znakiem. Pliki csv mogą również posiadać nagłówki kolumn, które nie różnią się niczym od pozostałych wierszy. Poniżej znajduje się przykład wczytania zbioru z pliku movies.csv przy jawnym wskazaniu parametrów charakteryzujących zawartość, takich jak separator w postaci przecinka, czy nagłówek w wierszu 0:

In [10]:
movie: pd.DataFrame = pd.read_csv('movie.csv', sep=',', header=0)  

Parametr sep wskazuje separator wartości w pliku csv, którego domyślną wartością jest przecinek. Argument header jest przeznaczony do wskazywania pozycji wiersza nagłówkowego w pliku csv. Jeżeli nagłówek nie istnieje, należy w parametrze header przekazać wartość None, a nazwy kolumn zostaną dopasowane automatycznie. 

### Reprezentacja kolumn oraz wierszy w ramce

W celu uzyskania listy nazw kolumn należy użyć atrybutu columns na obiekcie klasy DataFrame:

In [11]:
movie.columns  

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

W podobny sposób można uzyskać indeksy (lub nazwy) wierszy:

In [12]:
movie.index 

RangeIndex(start=0, stop=4916, step=1)

W przypadku zbioru filmów, wiersze są opatrzone indeksami numerycznymi w zakresie od 0 do 4915 włącznie. Istnieje jednak możliwość nadania dowolnych (w tym tekstowych) indeksów wierszom w ramce.

### Typy wartości w kolumnach ramki danych

W celu sprawdzenia typów danych, które znajdują się w poszczególnych kolumnach należy użyć atrybutu dtypes wywołanego na obiekcie DataFrame:

In [13]:
movie.dtypes  

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

Metoda value_counts jest przeznaczona do zliczania częstości wystąpień poszczególnych wartości. Można ją z powodzeniem zastosować w podsumowaniu częstości występowania typów danych w konkretnych kolumnach:

In [25]:
movie.dtypes.value_counts()  

float64    13
object     12
int64       3
Name: count, dtype: int64

### Parametry statystyczne kolumn w ramce

Za pomocą metody describe wywoływanej na obiekcie klasy DataFrame można uzyskać informacje o podstawowych parametrach statystycznych kolumn zawierających wartości numeryczne:

In [26]:
movie.describe()  

num_critic_for_reviews     duration  director_facebook_likes  \
count             4867.000000  4901.000000              4814.000000   
mean               137.988905   107.090798               691.014541   
std                120.239379    25.286015              2832.954125   
min                  1.000000     7.000000                 0.000000   
25%                 49.000000    93.000000                 7.000000   
50%                108.000000   103.000000                48.000000   
75%                191.000000   118.000000               189.750000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             4893.000000             4909.000000  4.054000e+03   
mean               631.276313             6494.488491  4.764451e+07   
std               1625.874802            15106.986884  6.737255e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                132.000000              607.000000  5.019656e+06   
50%                366.000000              982.000000  2.504396e+07   
75%                633.000000            11000.000000  6.110841e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     4.916000e+03                4916.000000           4903.000000   
mean      8.264492e+04                9579.815907              1.377320   
std       1.383222e+05               18164.316990              2.023826   
min       5.000000e+00                   0.000000              0.000000   
25%       8.361750e+03                1394.750000              0.000000   
50%       3.313250e+04                3049.000000              1.000000   
75%       9.377275e+04               13616.750000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           4895.000000  4.432000e+03  4810.000000   
mean             267.668846  3.654749e+07  2002.447609   
std              372.934839  1.002427e+08    12.453977   
min                1.000000  2.180000e+02  1916.000000   
25%               64.000000  6.000000e+06  1999.000000   
50%              153.000000  1.985000e+07  2005.000000   
75%              320.500000  4.300000e+07  2011.000000   
max             5060.000000  4.200000e+09  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             4903.000000  4916.000000   4590.000000           4916.000000  
mean              1621.923516     6.437429      2.222349           7348.294142  
std               4011.299523     1.127802      1.402940          19206.016458  
min                  0.000000     1.600000      1.180000              0.000000  
25%                277.000000     5.800000      1.850000              0.000000  
50%                593.000000     6.600000      2.350000            159.000000  
75%                912.000000     7.200000      2.350000           2000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

## Wybieranie kolumn

Istnieje kilka sposobów na wybór kolumn z ramki danych:

In [27]:
col: pd.Series = movie['genres']  

col

0       Action|Adventure|Fantasy|Sci-Fi
1              Action|Adventure|Fantasy
2             Action|Adventure|Thriller
3                       Action|Thriller
4                           Documentary
                     ...               
4911                       Comedy|Drama
4912       Crime|Drama|Mystery|Thriller
4913              Drama|Horror|Thriller
4914               Comedy|Drama|Romance
4915                        Documentary
Name: genres, Length: 4916, dtype: object

In [28]:
movie.genres 

0       Action|Adventure|Fantasy|Sci-Fi
1              Action|Adventure|Fantasy
2             Action|Adventure|Thriller
3                       Action|Thriller
4                           Documentary
                     ...               
4911                       Comedy|Drama
4912       Crime|Drama|Mystery|Thriller
4913              Drama|Horror|Thriller
4914               Comedy|Drama|Romance
4915                        Documentary
Name: genres, Length: 4916, dtype: object

Kolumny w ramkach danych opatrzone są typem Series, co oznacza że składają się z wartości opatrzonych indeksami wierszy z ramki. Oba sposoby wybierania kolumn zapewniają taki sam efekt, jednak użycie operatora kropki wyklucza wybór kolumny zawierającej w nazwie np. spację.

W podobny sposób można wybrać wiele kolumn z ramki danych przekazując listę ich nazw:

In [29]:
movie[['actor_1_name', 'actor_2_name', 'actor_3_name']].head()  

actor_1_name      actor_2_name          actor_3_name
0      CCH Pounder  Joel David Moore             Wes Studi
1      Johnny Depp     Orlando Bloom        Jack Davenport
2  Christoph Waltz      Rory Kinnear      Stephanie Sigman
3        Tom Hardy    Christian Bale  Joseph Gordon-Levitt
4      Doug Walker        Rob Walker                   NaN

### Wybór kolumn według typu danych

Kolumny z ramki można również wybrać na podstawie charakteru danych, które się w nich znajdują. Służy do tego metoda select_dtypes wywoływana na obiekcie klasy Dataframe.

In [30]:
movie.select_dtypes(include=['float']).head()  

num_critic_for_reviews  duration  director_facebook_likes  \
0                   723.0     178.0                      0.0   
1                   302.0     169.0                    563.0   
2                   602.0     148.0                      0.0   
3                   813.0     164.0                  22000.0   
4                     NaN       NaN                    131.0   

   actor_3_facebook_likes  actor_1_facebook_likes        gross  \
0                   855.0                  1000.0  760505847.0   
1                  1000.0                 40000.0  309404152.0   
2                   161.0                 11000.0  200074175.0   
3                 23000.0                 27000.0  448130642.0   
4                     NaN                   131.0          NaN   

   facenumber_in_poster  num_user_for_reviews       budget  title_year  \
0                   0.0                3054.0  237000000.0      2009.0   
1                   0.0                1238.0  300000000.0      2007.0   
2                   1.0                 994.0  245000000.0      2015.0   
3                   0.0                2701.0  250000000.0      2012.0   
4                   0.0                   NaN          NaN         NaN   

   actor_2_facebook_likes  imdb_score  aspect_ratio  
0                   936.0         7.9          1.78  
1                  5000.0         7.1          2.35  
2                   393.0         6.8          2.35  
3                 23000.0         8.5          2.35  
4                    12.0         7.1           NaN

### Wybór kolumn według frazy w nazwie

Metoda filter wywoływana na obiekcie klasy DataFrame zwróci ramkę danych z kolumnami, które zawierają wskazanę frazę za pomocą parametru like:

In [31]:
movie.filter(like='facebook').head()  

director_facebook_likes  actor_3_facebook_likes  actor_1_facebook_likes  \
0                      0.0                   855.0                  1000.0   
1                    563.0                  1000.0                 40000.0   
2                      0.0                   161.0                 11000.0   
3                  22000.0                 23000.0                 27000.0   
4                    131.0                     NaN                   131.0   

   cast_total_facebook_likes  actor_2_facebook_likes  movie_facebook_likes  
0                       4834                   936.0                 33000  
1                      48350                  5000.0                     0  
2                      11700                   393.0                 85000  
3                     106759                 23000.0                164000  
4                        143                    12.0                     0

## Praca z wartościami wybrakowanymi

Niektóre wartości w kolumnach ramki mogą być wybrakowane. Metoda isnull wywołana na obiekcie klasy Series zwróci serię wartości logicznych odpowiadających temu czy dana wartośc w kolumnie jest wybrakowana czy nie:

In [32]:
movie['genres'].isnull()  

0       False
1       False
2       False
3       False
4       False
        ...  
4911    False
4912    False
4913    False
4914    False
4915    False
Name: genres, Length: 4916, dtype: bool

Za pomocą metody any wywołanej na powstałej w ten sposób ramce można sprawdzić czy występuje tam przynajmniej jedna wartość prawdziwa:

In [33]:
movie['genres'].isnull().any()  

False

Metoda all umożliwia sprawdzenie czy w ramce występują tylko i wyłącznie wartości prawdziwe.

Jeżeli w kolumnie występują wartości wybrakowane, można je uzupełnić wskazaną wartością, np. zerem:

In [34]:
movie['actor_1_facebook_likes'].fillna(0)  

0        1000.0
1       40000.0
2       11000.0
3       27000.0
4         131.0
         ...   
4911      637.0
4912      841.0
4913        0.0
4914      946.0
4915       86.0
Name: actor_1_facebook_likes, Length: 4916, dtype: float64

Aby umieścić wartości wybrakowane w oryginalnej ramce danych należy użyć metody fillna na obiekcie klasy DataFrame przekazując jako parametr słownik mapujący nazwy kolumn na wartości, którymi mają zostać zastąpione wartości wybrakowane:

In [35]:
movie.fillna({  
    'actor_1_facebook_likes': 200,
    'actor_3_facebook_likes': 100,
})  

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
4911  Color        Scott Smith                     1.0      87.0   
4912  Color                NaN                    43.0      43.0   
4913  Color   Benjamin Roberds                    13.0      76.0   
4914  Color        Daniel Hsia                    14.0     100.0   
4915  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                   100.0        Rob Walker   
...                       ...                     ...               ...   
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
4911                   637.0          NaN                     Comedy|Drama   
4912                   841.0          NaN     Crime|Drama|Mystery|Thriller   
4913                     0.0          NaN            Drama|Horror|Thriller   
4914                   946.0      10443.0             Comedy|Drama|Romance   
4915                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
4911  ...                  6.0  English   Canada             NaN          NaN   
4912  ...                359.0  English      USA           TV-14          NaN   
4913  ...                  3.0  English      USA             NaN       1400.0   
4914  ...                  9.0  English      USA           PG-13          NaN   
4915  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

## Arytmetyka serii danych

Na obiektach klasy Series można stosować podstawowe operatory arytmetyczne. Przykładowo, do pięciu pierwszych wierszy kolumny imdb_score zostanie dodana wartość 1:

In [36]:
movie['imdb_score'].head()  

0    7.9
1    7.1
2    6.8
3    8.5
4    7.1
Name: imdb_score, dtype: float64

In [37]:
movie['imdb_score'].head() + 3 

0    10.9
1    10.1
2     9.8
3    11.5
4    10.1
Name: imdb_score, dtype: float64

W analogiczny sposób można dokonywać innych operacji arytmetycznych, takich jak odejmowanie, mnożenie, dzielenie.

## Operacje logiczne na serii danych

W wyniku operacji na ramkach danych, zostanie zwrócona nowa ramka zawierająca wartości logiczne będące wynikiem dokonanej operacji. Przykładowo, sprawdzenie czy 5 pierwszych wartości w kolumnie imdb_score są większe od 7.5:


In [38]:
movie['imdb_score'].head() > 7.5  

0     True
1    False
2    False
3     True
4    False
Name: imdb_score, dtype: bool

W analogiczny sposób można dokonywać operacji logicznych porównania, większości, mniejszości itd.

## Ustawianie indeksów w ramce danych

Metoda set_index umożliwia ustawienie wskazanej kolumny jako indeks ramki:

In [39]:
movie_new_idx: pd.DataFrame = movie.set_index('movie_title')  

movie_new_idx

color      director_name  \
movie_title                                                            
Avatar                                      Color      James Cameron   
Pirates of the Caribbean: At World's End    Color     Gore Verbinski   
Spectre                                     Color         Sam Mendes   
The Dark Knight Rises                       Color  Christopher Nolan   
Star Wars: Episode VII - The Force Awakens    NaN        Doug Walker   
...                                           ...                ...   
Signed Sealed Delivered                     Color        Scott Smith   
The Following                               Color                NaN   
A Plague So Pleasant                        Color   Benjamin Roberds   
Shanghai Calling                            Color        Daniel Hsia   
My Date with Drew                           Color           Jon Gunn   

                                            num_critic_for_reviews  duration  \
movie_title                                                                    
Avatar                                                       723.0     178.0   
Pirates of the Caribbean: At World's End                     302.0     169.0   
Spectre                                                      602.0     148.0   
The Dark Knight Rises                                        813.0     164.0   
Star Wars: Episode VII - The Force Awakens                     NaN       NaN   
...                                                            ...       ...   
Signed Sealed Delivered                                        1.0      87.0   
The Following                                                 43.0      43.0   
A Plague So Pleasant                                          13.0      76.0   
Shanghai Calling                                              14.0     100.0   
My Date with Drew                                             43.0      90.0   

                                            director_facebook_likes  \
movie_title                                                           
Avatar                                                          0.0   
Pirates of the Caribbean: At World's End                      563.0   
Spectre                                                         0.0   
The Dark Knight Rises                                       22000.0   
Star Wars: Episode VII - The Force Awakens                    131.0   
...                                                             ...   
Signed Sealed Delivered                                         2.0   
The Following                                                   NaN   
A Plague So Pleasant                                            0.0   
Shanghai Calling                                                0.0   
My Date with Drew                                              16.0   

                                            actor_3_facebook_likes  \
movie_title                                                          
Avatar                                                       855.0   
Pirates of the Caribbean: At World's End                    1000.0   
Spectre                                                      161.0   
The Dark Knight Rises                                      23000.0   
Star Wars: Episode VII - The Force Awakens                     NaN   
...                                                            ...   
Signed Sealed Delivered                                      318.0   
The Following                                                319.0   
A Plague So Pleasant                                           0.0   
Shanghai Calling                                             489.0   
My Date with Drew                                             16.0   

                                                actor_2_name  \
movie_title                                                    
Avatar                                      Joel David Moore   
Pirates of the Caribbean: At World's En

Taki sam efekt można uzyskać wskazując nazwę kolumny przy wczytywaniu pliku csv (funkcja read_csv) w parametrze index_col.

## Zmiana nazw kolumn oraz indeksów wierszy

W tym celu należy użyć metody rename wywołanej na obiekcie klasy DataFrame przekazując słowniki mapujące stare nazwy na nowe:

In [40]:
from typing import Dict

idx_rename: Dict[str, str] = {  
    'Avatar': 'Avatar II',
    'Spectre': 'Spectre II',
}

col_rename: Dict[str, str] = {  
    'director_facebook_likes': 'Director FB likes',
    'num_critic_for_reviews': 'Num critic for reviews',
}

movie_new_idx = movie_new_idx.rename(index=idx_rename, columns=col_rename)  

movie_new_idx.head(n=7)  

color      director_name  \
movie_title                                                            
Avatar II                                   Color      James Cameron   
Pirates of the Caribbean: At World's End    Color     Gore Verbinski   
Spectre II                                  Color         Sam Mendes   
The Dark Knight Rises                       Color  Christopher Nolan   
Star Wars: Episode VII - The Force Awakens    NaN        Doug Walker   
John Carter                                 Color     Andrew Stanton   
Spider-Man 3                                Color          Sam Raimi   

                                            Num critic for reviews  duration  \
movie_title                                                                    
Avatar II                                                    723.0     178.0   
Pirates of the Caribbean: At World's End                     302.0     169.0   
Spectre II                                                   602.0     148.0   
The Dark Knight Rises                                        813.0     164.0   
Star Wars: Episode VII - The Force Awakens                     NaN       NaN   
John Carter                                                  462.0     132.0   
Spider-Man 3                                                 392.0     156.0   

                                            Director FB likes  \
movie_title                                                     
Avatar II                                                 0.0   
Pirates of the Caribbean: At World's End                563.0   
Spectre II                                                0.0   
The Dark Knight Rises                                 22000.0   
Star Wars: Episode VII - The Force Awakens              131.0   
John Carter                                             475.0   
Spider-Man 3                                              0.0   

                                            actor_3_facebook_likes  \
movie_title                                                          
Avatar II                                                    855.0   
Pirates of the Caribbean: At World's End                    1000.0   
Spectre II                                                   161.0   
The Dark Knight Rises                                      23000.0   
Star Wars: Episode VII - The Force Awakens                     NaN   
John Carter                                                  530.0   
Spider-Man 3                                                4000.0   

                                                actor_2_name  \
movie_title                                                    
Avatar II                                   Joel David Moore   
Pirates of the Caribbean: At World's End       Orlando Bloom   
Spectre II                                      Rory Kinnear   
The Dark Knight Rises                         Christian Bale   
Star Wars: Episode VII - The Force Awakens        Rob Walker   
John Carter                                  Samantha Morton   
Spider-Man 3                                    James Franco   

                                            actor_1_facebook_likes  \
movie_title                                                          
Avatar II                                                   1000.0   
Pirates of the Caribbean: At World's End                   40000.0   
Spectre II                                                 11000.0   
The Dark Knight Rises                                      27000.0   
Star Wars: Episode VII - The Force Awakens                   131.0   
John Carter                                                  640.0   
Spider-Man 3                                               24000.0   

                                                  gross  \
movie_title                                               
Avatar II                                   760505847.0   
Pirates of the Caribbean: At World's End    309404152.0   
Spectre II   

## Tworzenie nowych kolumn

Kolumnę ze stałymi wartościami (np. 100) można dodać następująco: (kolumna znajdzie się na końcu ramki)

In [41]:
movie_new_idx['visits_num'] = 100  

movie_new_idx.head()

color      director_name  \
movie_title                                                            
Avatar II                                   Color      James Cameron   
Pirates of the Caribbean: At World's End    Color     Gore Verbinski   
Spectre II                                  Color         Sam Mendes   
The Dark Knight Rises                       Color  Christopher Nolan   
Star Wars: Episode VII - The Force Awakens    NaN        Doug Walker   

                                            Num critic for reviews  duration  \
movie_title                                                                    
Avatar II                                                    723.0     178.0   
Pirates of the Caribbean: At World's End                     302.0     169.0   
Spectre II                                                   602.0     148.0   
The Dark Knight Rises                                        813.0     164.0   
Star Wars: Episode VII - The Force Awakens                     NaN       NaN   

                                            Director FB likes  \
movie_title                                                     
Avatar II                                                 0.0   
Pirates of the Caribbean: At World's End                563.0   
Spectre II                                                0.0   
The Dark Knight Rises                                 22000.0   
Star Wars: Episode VII - The Force Awakens              131.0   

                                            actor_3_facebook_likes  \
movie_title                                                          
Avatar II                                                    855.0   
Pirates of the Caribbean: At World's End                    1000.0   
Spectre II                                                   161.0   
The Dark Knight Rises                                      23000.0   
Star Wars: Episode VII - The Force Awakens                     NaN   

                                                actor_2_name  \
movie_title                                                    
Avatar II                                   Joel David Moore   
Pirates of the Caribbean: At World's End       Orlando Bloom   
Spectre II                                      Rory Kinnear   
The Dark Knight Rises                         Christian Bale   
Star Wars: Episode VII - The Force Awakens        Rob Walker   

                                            actor_1_facebook_likes  \
movie_title                                                          
Avatar II                                                   1000.0   
Pirates of the Caribbean: At World's End                   40000.0   
Spectre II                                                 11000.0   
The Dark Knight Rises                                      27000.0   
Star Wars: Episode VII - The Force Awakens                   131.0   

                                                  gross  \
movie_title                                               
Avatar II                                   760505847.0   
Pirates of the Caribbean: At World's End    309404152.0   
Spectre II                                  200074175.0   
The Dark Knight Rises                       448130642.0   
Star Wars: Episode VII - The Force Awakens          NaN   

                                                                     genres  \
movie_title                                                                   
Avatar II                                   Action|Adventure|Fantasy|Sci-Fi   
Pirates of the Caribbean: At World's End           Action|Adventure|Fantasy   
Spectre II                                        Action|Adventure|Thriller   
The Dark Knight Rises                                       Action|Thriller   
Star Wars: Episode VII - The Force Awakens                      Documentary   

                                            ... language  country  \
movie_title                                 ...       

Biblioteka pandas umożliwia również tworzenie nowych kolumn w ramce na podstawie istniejących. Przykładowo, będzie to kolumna sumująca wszystkie lajki uzyskane przez aktorów i reżysera:

In [42]:
movie_new_idx['total_facebook_likes'] = (movie_new_idx['actor_1_facebook_likes'] + movie_new_idx['actor_2_facebook_likes'] + movie_new_idx['actor_3_facebook_likes'] + movie_new_idx['Director FB likes'])

movie_new_idx.head() 

color      director_name  \
movie_title                                                            
Avatar II                                   Color      James Cameron   
Pirates of the Caribbean: At World's End    Color     Gore Verbinski   
Spectre II                                  Color         Sam Mendes   
The Dark Knight Rises                       Color  Christopher Nolan   
Star Wars: Episode VII - The Force Awakens    NaN        Doug Walker   

                                            Num critic for reviews  duration  \
movie_title                                                                    
Avatar II                                                    723.0     178.0   
Pirates of the Caribbean: At World's End                     302.0     169.0   
Spectre II                                                   602.0     148.0   
The Dark Knight Rises                                        813.0     164.0   
Star Wars: Episode VII - The Force Awakens                     NaN       NaN   

                                            Director FB likes  \
movie_title                                                     
Avatar II                                                 0.0   
Pirates of the Caribbean: At World's End                563.0   
Spectre II                                                0.0   
The Dark Knight Rises                                 22000.0   
Star Wars: Episode VII - The Force Awakens              131.0   

                                            actor_3_facebook_likes  \
movie_title                                                          
Avatar II                                                    855.0   
Pirates of the Caribbean: At World's End                    1000.0   
Spectre II                                                   161.0   
The Dark Knight Rises                                      23000.0   
Star Wars: Episode VII - The Force Awakens                     NaN   

                                                actor_2_name  \
movie_title                                                    
Avatar II                                   Joel David Moore   
Pirates of the Caribbean: At World's End       Orlando Bloom   
Spectre II                                      Rory Kinnear   
The Dark Knight Rises                         Christian Bale   
Star Wars: Episode VII - The Force Awakens        Rob Walker   

                                            actor_1_facebook_likes  \
movie_title                                                          
Avatar II                                                   1000.0   
Pirates of the Caribbean: At World's End                   40000.0   
Spectre II                                                 11000.0   
The Dark Knight Rises                                      27000.0   
Star Wars: Episode VII - The Force Awakens                   131.0   

                                                  gross  \
movie_title                                               
Avatar II                                   760505847.0   
Pirates of the Caribbean: At World's End    309404152.0   
Spectre II                                  200074175.0   
The Dark Knight Rises                       448130642.0   
Star Wars: Episode VII - The Force Awakens          NaN   

                                                                     genres  \
movie_title                                                                   
Avatar II                                   Action|Adventure|Fantasy|Sci-Fi   
Pirates of the Caribbean: At World's End           Action|Adventure|Fantasy   
Spectre II                                        Action|Adventure|Thriller   
The Dark Knight Rises                                       Action|Thriller   
Star Wars: Episode VII - The Force Awakens                      Documentary   

                                            ... country  content_rating  \
movie_title                                 ... 

## Usuwanie kolumn

Do usuwania kolumn z ramki danych można wykorzystać instrukcję del:

In [43]:
del movie_new_idx['color']  

movie_new_idx.head()

director_name  \
movie_title                                                     
Avatar II                                       James Cameron   
Pirates of the Caribbean: At World's End       Gore Verbinski   
Spectre II                                         Sam Mendes   
The Dark Knight Rises                       Christopher Nolan   
Star Wars: Episode VII - The Force Awakens        Doug Walker   

                                            Num critic for reviews  duration  \
movie_title                                                                    
Avatar II                                                    723.0     178.0   
Pirates of the Caribbean: At World's End                     302.0     169.0   
Spectre II                                                   602.0     148.0   
The Dark Knight Rises                                        813.0     164.0   
Star Wars: Episode VII - The Force Awakens                     NaN       NaN   

                                            Director FB likes  \
movie_title                                                     
Avatar II                                                 0.0   
Pirates of the Caribbean: At World's End                563.0   
Spectre II                                                0.0   
The Dark Knight Rises                                 22000.0   
Star Wars: Episode VII - The Force Awakens              131.0   

                                            actor_3_facebook_likes  \
movie_title                                                          
Avatar II                                                    855.0   
Pirates of the Caribbean: At World's End                    1000.0   
Spectre II                                                   161.0   
The Dark Knight Rises                                      23000.0   
Star Wars: Episode VII - The Force Awakens                     NaN   

                                                actor_2_name  \
movie_title                                                    
Avatar II                                   Joel David Moore   
Pirates of the Caribbean: At World's End       Orlando Bloom   
Spectre II                                      Rory Kinnear   
The Dark Knight Rises                         Christian Bale   
Star Wars: Episode VII - The Force Awakens        Rob Walker   

                                            actor_1_facebook_likes  \
movie_title                                                          
Avatar II                                                   1000.0   
Pirates of the Caribbean: At World's End                   40000.0   
Spectre II                                                 11000.0   
The Dark Knight Rises                                      27000.0   
Star Wars: Episode VII - The Force Awakens                   131.0   

                                                  gross  \
movie_title                                               
Avatar II                                   760505847.0   
Pirates of the Caribbean: At World's End    309404152.0   
Spectre II                                  200074175.0   
The Dark Knight Rises                       448130642.0   
Star Wars: Episode VII - The Force Awakens          NaN   

                                                                     genres  \
movie_title                                                                   
Avatar II                                   Action|Adventure|Fantasy|Sci-Fi   
Pirates of the Caribbean: At World's End           Action|Adventure|Fantasy   
Spectre II                                        Action|Adventure|Thriller   
The Dark Knight Rises                                       Action|Thriller   
Star Wars: Episode VII - The Force Awakens                      Documentary   

                                               actor_1_name  ...  country  \
movie_title                                                  ...            
Avatar II             

## Wybór wierszy

Za pomocą atrybutu tablicowego loc ramki danych można wskazać wiersz lub wiersze do wyboru wskazując indeks:

In [44]:
movie_new_idx.loc['Avatar II']

director_name                                                    James Cameron
Num critic for reviews                                                   723.0
duration                                                                 178.0
Director FB likes                                                          0.0
actor_3_facebook_likes                                                   855.0
actor_2_name                                                  Joel David Moore
actor_1_facebook_likes                                                  1000.0
gross                                                              760505847.0
genres                                         Action|Adventure|Fantasy|Sci-Fi
actor_1_name                                                       CCH Pounder
num_voted_users                                                         886204
cast_total_facebook_likes                                                 4834
actor_3_name                                        

In [45]:
movie_new_idx.loc[['Avatar II', 'Spectre II']]  

director_name  Num critic for reviews  duration  \
movie_title                                                    
Avatar II    James Cameron                   723.0     178.0   
Spectre II      Sam Mendes                   602.0     148.0   

             Director FB likes  actor_3_facebook_likes      actor_2_name  \
movie_title                                                                
Avatar II                  0.0                   855.0  Joel David Moore   
Spectre II                 0.0                   161.0      Rory Kinnear   

             actor_1_facebook_likes        gross  \
movie_title                                        
Avatar II                    1000.0  760505847.0   
Spectre II                  11000.0  200074175.0   

                                      genres     actor_1_name  ...  country  \
movie_title                                                    ...            
Avatar II    Action|Adventure|Fantasy|Sci-Fi      CCH Pounder  ...      USA   
Spectre II         Action|Adventure|Thriller  Christoph Waltz  ...       UK   

             content_rating       budget  title_year actor_2_facebook_likes  \
movie_title                                                                   
Avatar II             PG-13  237000000.0      2009.0                  936.0   
Spectre II            PG-13  245000000.0      2015.0                  393.0   

            imdb_score  aspect_ratio movie_facebook_likes visits_num  \
movie_title                                                            
Avatar II          7.9          1.78                33000        100   
Spectre II         6.8          2.35                85000        100   

            total_facebook_likes  
movie_title                       
Avatar II                 2791.0  
Spectre II               11554.0  

[2 rows x 28 columns]

Jeżeli ramka danych posiada indeks numeryczny, w atrybucie loc wystarczy wskazać numeryczne pozycje wierszy:

In [46]:
movie.loc[[0, 1, 2]]  

color   director_name  num_critic_for_reviews  duration  \
0  Color   James Cameron                   723.0     178.0   
1  Color  Gore Verbinski                   302.0     169.0   
2  Color      Sam Mendes                   602.0     148.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  

[3 rows x 28 columns]

Atrybut tablicowy iloc (podobnie jak w bibliotece NumPy) służy do wybierania wierszy i kolumn wskazując ich pozycje numerycznie:

In [47]:
movie_new_idx.iloc[0:10, :]  

director_name  \
movie_title                                                     
Avatar II                                       James Cameron   
Pirates of the Caribbean: At World's End       Gore Verbinski   
Spectre II                                         Sam Mendes   
The Dark Knight Rises                       Christopher Nolan   
Star Wars: Episode VII - The Force Awakens        Doug Walker   
John Carter                                    Andrew Stanton   
Spider-Man 3                                        Sam Raimi   
Tangled                                          Nathan Greno   
Avengers: Age of Ultron                           Joss Whedon   
Harry Potter and the Half-Blood Prince            David Yates   

                                            Num critic for reviews  duration  \
movie_title                                                                    
Avatar II                                                    723.0     178.0   
Pirates of the Caribbean: At World's End                     302.0     169.0   
Spectre II                                                   602.0     148.0   
The Dark Knight Rises                                        813.0     164.0   
Star Wars: Episode VII - The Force Awakens                     NaN       NaN   
John Carter                                                  462.0     132.0   
Spider-Man 3                                                 392.0     156.0   
Tangled                                                      324.0     100.0   
Avengers: Age of Ultron                                      635.0     141.0   
Harry Potter and the Half-Blood Prince                       375.0     153.0   

                                            Director FB likes  \
movie_title                                                     
Avatar II                                                 0.0   
Pirates of the Caribbean: At World's End                563.0   
Spectre II                                                0.0   
The Dark Knight Rises                                 22000.0   
Star Wars: Episode VII - The Force Awakens              131.0   
John Carter                                             475.0   
Spider-Man 3                                              0.0   
Tangled                                                  15.0   
Avengers: Age of Ultron                                   0.0   
Harry Potter and the Half-Blood Prince                  282.0   

                                            actor_3_facebook_likes  \
movie_title                                                          
Avatar II                                                    855.0   
Pirates of the Caribbean: At World's End                    1000.0   
Spectre II                                                   161.0   
The Dark Knight Rises                                      23000.0   
Star Wars: Episode VII - The Force Awakens                     NaN   
John Carter                                                  530.0   
Spider-Man 3                                                4000.0   
Tangled                                                      284.0   
Avengers: Age of Ultron                                    19000.0   
Harry Potter and the Half-Blood Prince                     10000.0   

                                                 actor_2_name  \
movie_title                                                     
Avatar II                                    Joel David Moore   
Pirates of the Caribbean: At World's End        Orlando Bloom   
Spectre II                                       Rory Kinnear   
The Dark Knight Rises                          Christian Bale   
Star Wars: Episode VII - The Force Awakens         Rob Walker   
John Carter                                   Samantha Morton   
Spider-Man 3                                     James Franco   
Tangled                                          Donna Murphy   
Avengers: Age of Ultron           

In [48]:
movie_new_idx.iloc[:, 3:7]  

Director FB likes  \
movie_title                                                     
Avatar II                                                 0.0   
Pirates of the Caribbean: At World's End                563.0   
Spectre II                                                0.0   
The Dark Knight Rises                                 22000.0   
Star Wars: Episode VII - The Force Awakens              131.0   
...                                                       ...   
Signed Sealed Delivered                                   2.0   
The Following                                             NaN   
A Plague So Pleasant                                      0.0   
Shanghai Calling                                          0.0   
My Date with Drew                                        16.0   

                                            actor_3_facebook_likes  \
movie_title                                                          
Avatar II                                                    855.0   
Pirates of the Caribbean: At World's End                    1000.0   
Spectre II                                                   161.0   
The Dark Knight Rises                                      23000.0   
Star Wars: Episode VII - The Force Awakens                     NaN   
...                                                            ...   
Signed Sealed Delivered                                      318.0   
The Following                                                319.0   
A Plague So Pleasant                                           0.0   
Shanghai Calling                                             489.0   
My Date with Drew                                             16.0   

                                                actor_2_name  \
movie_title                                                    
Avatar II                                   Joel David Moore   
Pirates of the Caribbean: At World's End       Orlando Bloom   
Spectre II                                      Rory Kinnear   
The Dark Knight Rises                         Christian Bale   
Star Wars: Episode VII - The Force Awakens        Rob Walker   
...                                                      ...   
Signed Sealed Delivered                        Daphne Zuniga   
The Following                                  Valorie Curry   
A Plague So Pleasant                           Maxwell Moody   
Shanghai Calling                               Daniel Henney   
My Date with Drew                           Brian Herzlinger   

                                            actor_1_facebook_likes  
movie_title                                                         
Avatar II                                                   1000.0  
Pirates of the Caribbean: At World's End                   40000.0  
Spectre II                                                 11000.0  
The Dark Knight Rises                                      27000.0  
Star Wars: Episode VII - The Force Awakens                   131.0  
...                                                            ...  
Signed Sealed Delivered                                      637.0  
The Following                                                841.0  
A Plague So Pleasant                                           0.0  
Shanghai Calling                                             946.0  
My Date with Drew                                             86.0  

[4916 rows x 4 columns]

## Zadania

1. Wczytać zbiór danych z pliku movie_new.csv do ramki movie_new. Przy użyciu parametru names funkcji read_csv nadać następujące nazwy kolumnom: color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,actor_1_name,movie_title,num_voted_users,cast_total_facebook_likes,actor_3_name,facenumber_in_poster,plot_keywords,movie_imdb_link,num_user_for_reviews,language,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes.

2. Utworzyć ramkę movie_num, w której będą zawarte jedynie kolumny z wartościami typu numerycznego.

3. Wypełnić brakujące wartości w ramce movie_num średnią arytmetyczną wyznaczoną dla każdej kolumny z osobna. 

4. Znaleźć kolumnę w ramce movie_num, której wartości mają największe odchylenie standardowe.

5. Zastąpić kolumny numeryczne w ramce movie_new nowymi kolumnami z ramki movie_num.

6. Zastąpić wszystkie wartości wybrakowane w ramce movie_new za pomocą pierwszej wartości występującej w kolumnie.

7. Nadać nowe indeksy wierszom w ramce movie_new według następującego schematu: [imie i nazwisko rezysera] - [tytul filmu].

8. Zwiększyć wartość kolumny imdb_score każdego filmu w ramce movie_new o 1.

9. Zmienić nazwy wszystkich kolumn w ramce movie_new w taki sposób aby zlikwidować wszystkie znaki _ oraz pierwsze litery wszystkich słów w nazwach były wielkie.

10. Utworzyć ramkę movie_random, w której zostaną wybrane wiersze o losowej ilości w zakresie pozycji [start, end], gdzie start i end będą liczbami pseudolosowymi.

11. Zwiększyć wartości każdej kolumny numerycznej w ramce movie_random o wartość pseudolosową (po jednej wartości na kolumnę).